### ISSUE TITLE
Ability to use predicates (in the XPath sense) in property paths. e.g. "?doc :hasKeyword[dcterms:source :database1]/rdf:value ?keyword"

### ISSUE THREAD

**Issue by tfrancart** [w3c/sparql-12#99]
## Why?

The more detailled the graphs we make, the more intermediate nodes we have.
The closest we want to get to property graphs, the more intermediate nodes we create, e.g. to reify properties.
We thus need a convenient way to navigate these reified relations. I recently learned a way to achieve this in OWL is to use [rolification](https://stackoverflow.com/questions/16989042/owl-2-rolification), the ability to turn a class into an equivalent object property, in order to use it in property chain axioms. What if we had something equivalent in SPARQL, as an ability to include conditions on the node traversed by a property graph ?

## Previous work

[XPath predicates](https://www.w3.org/TR/1999/REC-xpath-19991116/#predicates)

## Proposed solution

```
  # select the minister in charge of the ministry in which a person works
  ?person ^org:hasMember[rdf:type :Ministry]/^org:headOf ?minister
```

would be equivalent to 

```
  ?person ^org:hasMember ?organisation .
  ?organisation rdf:type :Ministry .
  ?organisation ^org:headOf ?minister .
```

This could be a nice way to filter on provenance, quality metadata, etc.

```
  # select the keyword of a document coming from :database 1
  ?doc :hasKeyword[dcterms:source :database1]/rdf:value ?keyword
```

equivalent to

```
  ?doc :hasKeyword ?x .
  ?x dcterms:source :database1 .
  ?x rdf:value ?keyword .
```

In a sense this is simple synctactic sugar, but it allows nice filtering on reified properties. The easier this navigation is, the easier it would be to manage complex graph structures

## Considerations for backward compatibility

A syntax using `[...]` may conflict with the now lost syntax that allowed to filter on a certain number of recursive relations e.g. `skos:narrower|1,3]`

**Comment by VladimirAlexiev** [comment]
I like this proposal. 

While OWL2 has powerful class constructs, its property constructs are weak.
Being able to branch off a prop chain and check something on the side is useful.
I wrote something similar in 2014: http://rawgit2.com/VladimirAlexiev/my/master/pubs/extending-owl2/index.html search for "Chain and typecheck".

@tfrancart the "number of repetitions" syntax is not lost, see `elt{n,m}` in https://www.w3.org/TR/sparql11-property-paths/.

The proposed syntax is nice since it meshes with the syntax for blank nodes:
```sparql
?person ^org:hasMember[rdf:type :Ministry]
```
is perfectly valid in SPARQL 1.1. So you're proposing to be able to extend a path from this point onward.

However, rather than just referring to XPath predicates, we need to flesh out the syntax.
- using more props inside the brackets will be nice
- using vars inside the brackets to find bindings will be nice
- using filters inside the brackets: I don't like that. Filter outside of the patterns

**Comment by tfrancart** [comment]
> @tfrancart the "number of repetitions" syntax is not lost, see elt{n,m} in https://www.w3.org/TR/sparql11-property-paths/.

Ha yes, I could not remember the exact syntax

> So you're proposing to be able to extend a path from this point onward.

As you wrote, the idea is exactly to _branch off a prop chain and check something on the side_

> However, rather than just referring to XPath predicates, we need to flesh out the syntax.

Yes, this is more of an _intuition_ for the moment.

> using more props inside the brackets will be nice

Could be as simple as :

```
?doc :hasKeyword[dcterms:source :database1; dcterms:creator :Francis]/rdf:value ?keyword
```

> using vars inside the brackets to find bindings will be nice

Something like this ?

```
?doc :hasKeyword[dcterms:creator ?creator]/rdf:value ?keyword
```

> using filters inside the brackets: I don't like that. Filter outside of the patterns

Let's dream a bit, we are not _exactly_ in a triple pattern, after all :

```
?doc :hasKeyword[dcterms:created < "2019-06-01^^xsd:date"]/rdf:value ?keyword
```

**Comment by PaulWarrenOU** [comment]
This seems to me like a simple case of the various proposals for extended property paths, e.g. 
* Pérez, Jorge, Marcelo Arenas, and Claudio Gutierrez. “nSPARQL: A Navigational Language for RDF.” Web Semantics: Science, Services and Agents on the World Wide Web 8, no. 4 (2010): 255–270. 
* Fionda, Valeria, Giuseppe Pirrò, and Mariano P. Consens. “Querying Knowledge Graphs with Extended Property Paths.” Semantic Web, (not yet accepted: http://www.semantic-web-journal.net/content/querying-knowledge-graphs-extended-property-paths).
These and other authors have done a lot of work on the computational tractability of these kind of extensions.

**Comment by jaw111** [comment]
It's worth to mention [LDPath](https://marmotta.apache.org/ldpath/language.html) from Apache Marmotta as previous work here.

**Comment by PaulWarrenOU** [comment]
Yes, it was not my intention to be exhaustive, merely to point out that a lot of work has already been done in this area.

**Comment by jaw111** [comment]
I run into a use case now where this would be useful in combination with arbitrary length path matches.

Something like:

```
?minister (org:headOf [ a :Ministry ] / org:hasMember)+ ?person
```
